### PR TITLE
systemd: add missing TPM2 build dependencies

### LINKED ISSUE #139433: Expand support for and document systemd-cryptenroll
### Describe the issue

The upgrade from systemd 247 to 249 recently landed in nixpkgs-unstable. Version 248 brought support for `systemd-cryptenroll`, which allows for automatica unlocking of LUKS2 volumes with diverse security hardware.

In my specific case, I would like to use a FIDO2 security key as described in Lennart Poettering's [blog post](http://0pointer.net/blog/unlocking-luks2-volumes-with-tpm2-fido2-pkcs11-security-hardware-on-systemd-248.html), which requires an entry in `/etc/crypttab`.

Running `systemd-cryptenroll` correctly configures LUKS slots and tokens, and setting

```nix
environment.etc.crypttab = {
    enable = true;
    text = ''
      cryptroot /dev/sda2 - fido2-device=auto
    '';
  };
```

in my system configuration provides the correct file and appears to integrate with systemd when rebuilding. However I am still asked to provide a password for `/dev/sda2` when rebooting with zero interaction from my security key.

### Expected behavior

Unlocking LUKS at boot by confirming user presence. Ideally this could be managed by a NixOS Module such as [luksroot](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/luksroot.nix).

### Additional Context

- I am using a SoloKeys Somu, which supports U2F and FIDO2

- I believe this could be related to https://github.com/NixOS/nixpkgs/issues/31258.

### Notify maintainers

@andir @eelco @flokli @kloenk

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```yaml
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.68, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210908_3c56f62`
 - channels(root): `"nixos-21.11pre313466.bc06c93905f"`
 - channels(johannes): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: cryptsetup
# a list of nixos modules affected by the problem
module: systemd
```

### PR DESCRIPTION
**Pull request by ymatsiuk**
###### Motivation for this change
before:
```
result/bin/systemd-cryptenroll --tpm2-device=list
TPM2 not supported on this build.
```
after:
```
result/bin/systemd-cryptenroll --tpm2-device=list
PATH        DEVICE     DRIVER
/dev/tpmrm0 STM0125:00 tpm_tis
```

Fixes #139433.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Kloenk** [comment]
This should target `staging` as it's a mass rebuild

**Comment by ymatsiuk** [comment]
Tests will follow soon, finalizing some minor issues

**Comment by ymatsiuk** [comment]
tagging @cwyc and @flokli from #134577

File: nixos/tests/systemd-cryptenroll.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    machine.succeed("test -e /dev/tpm0")
+    machine.succeed("test -e /dev/tpmrm0")
+    machine.succeed("systemd-cryptenroll --tpm2-device=list")
+    machine.succeed("echo -n lukspass | cryptsetup luksFormat -q /dev/vdb -")
+    machine.succeed("PASSWORD=lukspass systemd-cryptenroll --tpm2-device=auto --tpm2-pcrs=7 /dev/vdb")
+    machine.fail("PASSWORD=wrongpass systemd-cryptenroll --tpm2-device=auto --tpm2-pcrs=7 /dev/vdb")
+    machine.succeed("${pkgs.systemd}/lib/systemd/systemd-cryptsetup attach luks /dev/vdb - tpm2-device=auto")
+    machine.succeed("test -e /dev/mapper/luks")
+    machine.succeed("${pkgs.systemd}/lib/systemd/systemd-cryptsetup detach luks")
+    machine.succeed("systemd-cryptenroll --wipe-slot=tpm2 /dev/vdb")
```
Comment (flokli) [inline nixos/tests/systemd-cryptenroll.nix:40]: Can we add this (or another partition) to `/etc/crypttab`, and see if unlocking via that works? (similar to nixos/tests/systemd.nix, L146)

File: nixos/tests/systemd-cryptenroll.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
+let
```
Comment (flokli) [inline nixos/tests/systemd-cryptenroll.nix:1]: Please add this file to `nixos/tests/all-tests.nix` as well.

**Review by flokli — COMMENTED** [review]
Currently running the build, will report back. Some nits.

**Comment by flokli** [comment]
Tests succeeded :tada: So we should probably just extend `/etc/crypttab` setup and add this to `all-tests.nix`.

**Comment by ymatsiuk** [comment]
Thanks for the review @flokli

**Review by flokli — CHANGES_REQUESTED** [review]

**Comment by ymatsiuk** [comment]
I've separated the test to logical blocks and added comments for the context

**Review by flokli — APPROVED** [review]
LGTM. Thanks a lot for being patient with my nitpicks ;-)

**Pull request merged**